snap/quota/quota.go

```diff
@@ -153,6 +221,267 @@ func (grp *Group) SliceFileName() string {
 	return buf.String()
 }
 
+// groupQuotaAllocations contains information about current quotas of a group
+// and is used by getgroupQuotaAllocations to contain this information.
+// There are two types of values for each quota - the quota limit set by this group,
+// and the quota reserved by children of this group. Examples:
+// Group that has a non-memory quota, but has a child group that has a memory quota of 512mb:
+// memoryLimit = 0
+// memoryReserved = 512 mb
+// Group that has a memory quota of 512mb, but has only children groups with non-memory quota:
+// memoryLimit = 512 mb
+// memoryReserved = 0
+// Group that has a memory quota of 512mb, and has a child group that has a memory quota of 256mb:
+// memoryLimit = 512 mb
+// memoryReserved = 256 mb
+// If the limit value is non-zero, then the reserved value can never be greater than the limit, however
+// if the limit is zero, then the reserved value must be below the nearest non-zero limit as you traverse
+// up the tree.
+type GroupQuotaAllocations struct {
+	MemoryLimit    quantity.Size
+	MemoryReserved quantity.Size
+
+	CpuLimit    int
+	CpuReserved int
+
+	ThreadsLimit    int
+	ThreadsReserved int
+
+	AllowedCPUsLimit    []int
+	AllowedCPUsReserved []int
+}
+
+func max(a, b int) int {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func maxq(a, b quantity.Size) quantity.Size {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func (grp *Group) getAllowedCPUs() []int {
+	if grp.CPULimit == nil || len(grp.CPULimit.AllowedCPUs) == 0 {
+		return []int{}
+	}
+	return grp.CPULimit.AllowedCPUs
+}
+
+func (grp *Group) getTotalCPUPercentage() int {
+	if grp.CPULimit == nil || grp.CPULimit.Percentage == 0 {
+		return 0
+	}
+	return max(grp.CPULimit.Count, 1) * grp.CPULimit.Percentage
+}
+
+// getgroupQuotaAllocations Recursively retrieve current group quotas statistics, this should just
+// be invoked on the upper parent of a group tree, and then it will gather active quotas for the
+// tree and store them in the allQuotas paramater
+func (grp *Group) getgroupQuotaAllocations(allQuotas map[string]*GroupQuotaAllocations, skipGrp *Group) *GroupQuotaAllocations {
```

Comment:
this is a bit of an unusual name, it could be probably be called just getQuotaAllocations as it lives on the Group

Reply:
Yea something definitely went wrong with my renaming, I'll fix it